packages/flutter_tools/lib/src/test/flutter_platform.dart

```diff
@@ -902,23 +902,48 @@ class _FlutterPlatform extends PlatformPlugin {
   }
 }
 
+// The [_shellProcessClosed] future can't have errors thrown on it because it
+// crosses zones (it's fed in a zone created by the test package, but listened
+// to by a parent zone, the same zone that calls [close] below).
+//
+// This is because Dart won't let errors that were fed into a Future in one zone
+// propagate to listeners in another zone. (Specifically, the zone in which the
+// future was completed with the error, and the zone in which the listener was
+// registered, are what matters.)
+//
+// Because of this, the [_shellProcessClosed] future takes an [_AsyncError]
+// object as a result. If it's null, it's as if it had completed correctly; if
+// it's non-null, it contains the error and stack trace of the actual error, as
+// if it had completed with that error.
 class _FlutterPlatformStreamSinkWrapper<S> implements StreamSink<S> {
   _FlutterPlatformStreamSinkWrapper(this._parent, this._shellProcessClosed);
+
   final StreamSink<S> _parent;
-  final Future<void> _shellProcessClosed;
+  final Future<_AsyncError> _shellProcessClosed;
 
   @override
   Future<void> get done => _done.future;
   final Completer<void> _done = new Completer<void>();
 
   @override
   Future<dynamic> close() {
-   Future.wait<dynamic>(<Future<dynamic>>[
+    Future.wait<dynamic>(<Future<dynamic>>[
       _parent.close(),
       _shellProcessClosed,
     ]).then<void>(
-      (List<dynamic> value) {
-        _done.complete();
+      (List<dynamic> results) {
+        final dynamic result = results.reduce((dynamic result, dynamic value) {
+          if (value is _AsyncError) {
+            assert(result == null);
+            result = value;
+          }
+          return result;
+        });
+        if (result is _AsyncError) {
```

Comment:
The `results.reduce` above could be removed, and this could be simplified to `if (results[1] is _AsyncError)`

Reply:
The documentation for `Future.wait` doesn't say anything about the order of the values as far as I can tell. Is the order guaranteed? I'll ask the Dart folks...

Reply:
Dart folks say yes, the order should match. I'll remove the reduce.

Reply:
(Filed https://github.com/dart-lang/sdk/issues/34290 to improve docs.)